src/main/java/org/opensearch/neuralsearch/plugin/query/NeuralQueryBuilder.java

```diff
@@ -172,10 +193,32 @@ private static void parseQueryParams(XContentParser parser, NeuralQueryBuilder n
         }
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (vectorSupplier() != null && vectorSupplier.get() != null) {
+            return new KNNQueryBuilder(fieldName(), vectorSupplier.get(), k());
+        }
+
+        if (vectorSupplier() == null) {
+            SetOnce<float[]> vectorSetOnce = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                ((client, actionListener) -> ML_CLIENT.inferenceSentence(modelId(), queryText(), ActionListener.wrap(floatList -> {
+                    vectorSetOnce.set(vectorAsListToArray(floatList));
+                    actionListener.onResponse(null);
```

Comment:
My understanding here: doRewrite method will be invoked by [rewriteAndFetch](https://github.com/opensearch-project/OpenSearch/blob/e44158d4d10d4f8905895ffa50bf9398b8550667/server/src/main/java/org/opensearch/index/query/Rewriteable.java#L109) which is a recursive method by wrapping itself invocation in the listener. And once the predict invocation is done, the vectorSupplier will be populated with valid data, and a KNNQueryBuilder is returned, next recursive call will hit the check on vectorSupplier and return KNNQueryBuilder again, then this method stop. 
When the action gets invoked at [executeAsyncActions](https://github.com/opensearch-project/OpenSearch/blob/e44158d4d10d4f8905895ffa50bf9398b8550667/server/src/main/java/org/opensearch/index/query/QueryRewriteContext.java#L152), the `actionListener.onResponse(null)` here means to invoke the [wildcardListener's on Response](https://github.com/opensearch-project/OpenSearch/blob/e44158d4d10d4f8905895ffa50bf9398b8550667/server/src/main/java/org/opensearch/index/query/QueryRewriteContext.java#L135) to count down.

Reply:
Right, so we need to call listener on null so countdown happens

Reply:
@jmazanec15 please add these code links as documentation on top of this code.